cgp/genome.py

```diff
@@ -267,6 +267,184 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        used_node_indices: List[int] = []
+
+        while len(addable_nodes) > 0:
+
+            current_node_idx = rng.choice(list(addable_nodes))
+
+            dna = self._copy_dna_segment(
+                dna, idx_old_location=current_node_idx, idx_new_location=node_idx
+            )
+
+            for dependencies in node_dependencies.values():
+                dependencies.discard(current_node_idx)
+
+            [addable_nodes, used_node_indices] = self._update_addable_nodes(
+                node_dependencies, used_node_indices, current_idx=current_node_idx
+            )
+            node_idx += 1
+
+        self._update_gene_values(dna, used_node_indices)
+        self._replace_invalid_gene_values(dna, rng)
+
+        self.dna = dna
+
+    def _copy_dna_segment(
+        self, dna: List[int], idx_old_location: int, idx_new_location: int
+    ) -> List[int]:
+        """ Moves a nodes dna from its old node location to a new location. """
+
+        node_dna = self._dna[
+            idx_old_location
+            * self._length_per_region : (idx_old_location + 1)
+            * self._length_per_region
+        ]
+        dna[
+            idx_new_location
+            * self._length_per_region : (idx_new_location + 1)
+            * self._length_per_region
+        ] = node_dna
+
+        return dna
+
+    def _update_addable_nodes(
+        self, node_dependencies: Dict, used_node_indices: List[int], current_idx: int
+    ) -> [set, List[int]]:
+        """ Update the set of addable nodes,
+        by reevaluating which nodes haves no dependencies
+        and remove nodes which were already placed from the list.
+        """
+        addable_nodes = self._get_addable_nodes(node_dependencies)
+        used_node_indices.append(current_idx)
+        addable_nodes = addable_nodes.difference(used_node_indices)
+
+        return addable_nodes, used_node_indices
+
+    def _update_gene_values(self, dna: List[int], used_node_indices: List[int]) -> None:
+        """ Update gene values to correspond to the new position of the corresponding node"""
+        for gene_idx, gene_value in enumerate(dna):
+            region_idx = self._get_region_idx(gene_idx)
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                if gene_value >= self._n_inputs:
+                    gene_value = self._n_inputs + used_node_indices.index(gene_value)
+            dna[gene_idx] = gene_value
+
+    def _replace_invalid_gene_values(
+        self, dna_for_reorder: List[int], rng: np.random.RandomState
+    ) -> None:
+        """ Replace gene values of unused arities
+        by (pseudo-) random picks of permissible values for that node.
+        Works only in self.n_rows==1
+        """
+        if not self._n_rows == 1:
+            raise ValueError("Replacing invalid gene values only implemented for n_rows = 1")
```

Comment:
isn't this already caught at a higher level? would this ever evaluate to true? if it's just for reminding us that this holds only for `n_rows = 1`, then i would prefer just a comment in the code